src/imitation/data/types.py

```diff
@@ -22,6 +27,177 @@
+        if not isinstance(other, self.__class__):
+            return False
+        if not self.d.keys() == other.d.keys():
+            return False
+        return all(np.array_equal(self.d[k], other.d[k]) for k in self.d.keys())
+
+    @property
+    def shape(self) -> Dict[str, Tuple[int, ...]]:
+        """Returns a dictionary with shape-tuples in place of the arrays."""
+        return {k: v.shape for k, v in self.d.items()}
+
+    @property
+    def dtype(self) -> Dict[str, np.dtype]:
+        """Returns a dictionary with shape-tuples in place of the arrays."""
+        return {k: v.dtype for k, v in self.d.items()}
+
+    def unwrap(self) -> Dict[str, np.ndarray]:
+        return self.d
+
+    @classmethod
+    def maybe_wrap(
+        cls,
+        obs: Union[Dict[str, np.ndarray], np.ndarray, "DictObs"],
+    ) -> Union["DictObs", np.ndarray]:
+        """Converts an observation into a DictObs, if necessary."""
+        if isinstance(obs, dict):
+            return cls(obs)
+        else:
+            assert isinstance(obs, (np.ndarray, cls))
+            return obs
+
+    TensorVar = TypeVar("TensorVar", np.ndarray, th.Tensor)
+
+    @overload
+    @classmethod
+    def maybe_unwrap(cls, maybe_dictobs: "DictObs") -> Dict[str, np.ndarray]:
+        ...
+
+    @overload
+    @classmethod
+    def maybe_unwrap(cls, maybe_dictobs: TensorVar) -> TensorVar:
+        ...
+
+    @classmethod
+    def maybe_unwrap(cls, maybe_dictobs):
+        if isinstance(maybe_dictobs, cls):
+            return maybe_dictobs.unwrap()
+        else:
+            if not isinstance(maybe_dictobs, (np.ndarray, th.Tensor)):
+                warnings.warn(f"trying to unwrap object of type {type(maybe_dictobs)}")
+            return maybe_dictobs
+
+    def map_arrays(self, fn: Callable[[np.ndarray], np.ndarray]) -> "DictObs":
+        return self.__class__({k: fn(v) for k, v in self.d.items()})
+
+    @staticmethod
+    def _unravel(dictobs_list: Iterable["DictObs"]) -> Dict[str, List[np.ndarray]]:
+        """Converts a list of DictObs into a dictionary of lists of arrays."""
+        unraveled: Dict[str, List[np.ndarray]] = collections.defaultdict(list)
+        for do in dictobs_list:
+            for k, array in do.d.items():
+                unraveled[k].append(array)
+        return unraveled
+
+    @classmethod
+    def stack(cls, dictobs_list: Iterable["DictObs"]) -> "DictObs":
+        return cls(
+            {
+                k: np.stack(arr_list)
+                for k, arr_list in cls._unravel(dictobs_list).items()
+            },
+        )
+
+    @classmethod
+    def concatenate(cls, dictobs_list: Iterable["DictObs"]) -> "DictObs":
+        return cls(
+            {
+                k: np.concatenate(arr_list)
+                for k, arr_list in cls._unravel(dictobs_list).items()
+            },
+        )
+
+    # TODO: add keys, values, items?
+
+
+def assert_not_dictobs(x: Union[np.ndarray, DictObs]) -> np.ndarray:
```

Comment:
Nit: `assert` sounds like it should raise an AssertionError but this raises a ValueError. Consider changing name of method or the exception raised.